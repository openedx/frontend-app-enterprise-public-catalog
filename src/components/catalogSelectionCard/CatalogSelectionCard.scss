.catalog-selection-card-checkbox {
  display: flex;
  flex-direction: row-reverse;
  justify-content: space-between;
  overflow-wrap: anywhere; /* this allows word breaking to avoid weird layout in certain sizes */

  & .pgn__form-radio-input {
    /* disables the dynamic resizing of the radio boxes */
    flex-grow: 0;
    flex-shrink: 0;
  }
}
