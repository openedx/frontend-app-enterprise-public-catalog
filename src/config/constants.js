export const FEATURE_ENABLE_PROGRAMS = 'ENABLE_PROGRAMS';
export const FEATURE_PROGRAM_TYPE_FACET = 'ENABLE_PROGRAM_TYPE_FACET';
export const FEATURE_EXEC_ED_INCLUSION = 'ENABLE_EXEC_ED_INCLUSION';
