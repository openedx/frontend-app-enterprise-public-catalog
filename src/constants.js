/* eslint-disable import/prefer-default-export */
export const PAGE_TITLE = 'edX Enterprise Catalogs';

export const NUM_RESULTS_PER_PAGE = 40;

/**
 * Event names to use for tracking
 * Spec: https://openedx.atlassian.net/wiki/spaces/AN/pages/2624455676/Enterprise+public+catalog+Event+Design
 */
export const PRODUCT_EVENTS = {
  ctaButtonPressed: 'edx.enterprise.explore-catalog.cta_button_pressed',
};
export const TRACKING_APP_NAME = 'explore-catalog';
// end: tracking related

export const QUERY_UUID_REFINEMENT = 'enterprise_catalog_query_uuids';
export const AVAILABILITY_REFINEMENT = 'availability';
export const AVAILABILITY_REFINEMENT_DEFAULTS = 'Available Now,Upcoming';
