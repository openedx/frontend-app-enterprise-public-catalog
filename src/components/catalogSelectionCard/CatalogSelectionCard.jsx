import React, {
  useContext,
} from 'react';
import PropTypes from 'prop-types';
import {
  SearchContext, setRefinementAction,
} from '@edx/frontend-enterprise-catalog-search';
import {
  Badge, Card, Form,
} from '@edx/paragon';
import { QUERY_TITLE_REFINEMENT } from '../../constants';

export const CardCheckbox = ({
  label, queryTitle, labelDetail,
}) => {
  const { refinements, dispatch } = useContext(SearchContext);
  const isChecked = refinements[QUERY_TITLE_REFINEMENT]?.includes(queryTitle) || false;

  const setChecked = () => {
    if (!isChecked) {
      dispatch(setRefinementAction(QUERY_TITLE_REFINEMENT, [queryTitle]));
    }
  };
  return (
    <Form.Radio className="catalog-selection-card" checked={isChecked} onChange={setChecked} description={labelDetail}>{label}</Form.Radio>
  );
};

CardCheckbox.propTypes = {
  queryTitle: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  labelDetail: PropTypes.string.isRequired,
};

const CatalogSelectionCard = ({
  queryTitle, className, badgeVariant, badge, label, cardBody, labelDetail,
}) => (
  <Card>
    <Card.Body>
      <Card.Title>
        <Badge className={className} variant={badgeVariant}>
          {badge}
        </Badge>
        <CardCheckbox
          label={label}
          labelDetail={labelDetail}
          queryTitle={queryTitle}
        />
      </Card.Title>
      <Card.Text>{cardBody}</Card.Text>
    </Card.Body>
  </Card>
);

CatalogSelectionCard.defaultProps = {
  className: '',
};

CatalogSelectionCard.propTypes = {
  className: PropTypes.string,
  badgeVariant: PropTypes.string.isRequired,
  badge: PropTypes.string.isRequired,
  cardBody: PropTypes.string.isRequired,
  queryTitle: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  labelDetail: PropTypes.string.isRequired,
};

export default CatalogSelectionCard;
