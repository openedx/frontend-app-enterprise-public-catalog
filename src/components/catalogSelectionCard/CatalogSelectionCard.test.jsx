import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import React from 'react';
import {
  SEARCH_FACET_FILTERS,
  SearchContext,
  setRefinementAction,
} from '@edx/frontend-enterprise-catalog-search';
import { render, screen } from '@testing-library/react';
import { CardCheckbox } from './CatalogSelectionCard';
import { QUERY_UUID_REFINEMENT } from '../../constants';

const dispatchSpy = jest.fn();
const DEFAULT_SEARCH_CONTEXT_VALUE = { refinementsFromQueryParams: {}, dispatch: dispatchSpy };

// eslint-disable-next-line react/prop-types
const SearchDataWrapper = ({ children, searchContextValue }) => (
  <SearchContext.Provider
    value={searchContextValue}
    searchFacetFilters={
        [...SEARCH_FACET_FILTERS, { attribute: QUERY_UUID_REFINEMENT, title: 'Uuids' }]
      }
  >
    {children}
  </SearchContext.Provider>
);

describe('CardCheckbox', () => {
  beforeEach(() => {
    dispatchSpy.mockReset();
  });
  it('renders a checkbox and label', () => {
    const tree = renderer.create(
      <SearchDataWrapper
        searchContextValue={DEFAULT_SEARCH_CONTEXT_VALUE}
      >
        <CardCheckbox
          label="foo"
          queryUuid="bar"
          labelDetail="baz"
        />
      </SearchDataWrapper>,
    );
    expect(tree).toMatchSnapshot();
  });
  it('is checked when uuid is included in the query param', () => {
    const queryUuid = 'bar';
    const refinements = { refinementsFromQueryParams: { [QUERY_UUID_REFINEMENT]: [queryUuid] } };
    const tree = renderer.create(
      <SearchDataWrapper
        searchContextValue={refinements}
      >
        <CardCheckbox
          label="foo"
          queryUuid={queryUuid}
          labelDetail="baz"
        />
      </SearchDataWrapper>,
    );
    expect(tree).toMatchSnapshot();
  });
  it('dispatches setRefinement action when radio selected', () => {
    const label = 'Check me';
    const queryUuid = 'idforyou';
    render(
      <SearchDataWrapper searchContextValue={DEFAULT_SEARCH_CONTEXT_VALUE}>
        <CardCheckbox
          label={label}
          queryUuid={queryUuid}
          labelDetail="baz"
        />
      </SearchDataWrapper>,
    );
    const input = screen.getByLabelText(label);
    userEvent.click(input);
    expect(dispatchSpy).toHaveBeenLastCalledWith(setRefinementAction(QUERY_UUID_REFINEMENT, [queryUuid]));
  });
});
