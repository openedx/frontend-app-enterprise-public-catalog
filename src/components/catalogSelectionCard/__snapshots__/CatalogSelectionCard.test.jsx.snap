// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardCheckbox is checked when title is included in the query param 1`] = `
<div
  className="pgn__form-radio catalog-selection-card"
>
  <input
    aria-describedby="form-field4-6"
    checked={true}
    className="pgn__form-radio-input"
    id="form-field4"
    onChange={[Function]}
    type="radio"
  />
  <div>
    <label
      className="pgn__form-label"
      htmlFor="form-field4"
    >
      foo
    </label>
    <div
      className="pgn__form-control-description pgn__form-text pgn__form-text-default"
      id="form-field4-6"
    >
      baz
    </div>
  </div>
</div>
`;

exports[`CardCheckbox renders a checkbox and label 1`] = `
<div
  className="pgn__form-radio catalog-selection-card"
>
  <input
    checked={false}
    className="pgn__form-radio-input"
    id="form-field1"
    onChange={[Function]}
    type="radio"
  />
  <div>
    <label
      className="pgn__form-label"
      htmlFor="form-field1"
    >
      foo
    </label>
    <div
      className="pgn__form-control-description pgn__form-text pgn__form-text-default"
    >
      baz
    </div>
  </div>
</div>
`;
